usage/src/org/apache/cloudstack/quota/QuotaAlertManagerImpl.java

```diff
@@ -0,0 +1,471 @@
+            s_logger.info("checkAndSendQuotaAlertEmails accId=" + quotaAccount.getId());
+            BigDecimal accountBalance = quotaAccount.getQuotaBalance();
+            Date balanceDate = quotaAccount.getQuotaBalanceDate();
+            Date alertDate = quotaAccount.getQuotaAlertDate();
+            int lockable = quotaAccount.getQuotaEnforce();
+            BigDecimal thresholdBalance = quotaAccount.getQuotaMinBalance();
+            if (accountBalance != null) {
+                AccountVO account = _accountDao.findById(quotaAccount.getId());
+                // s_logger.info("Check id " + account.getId() + " bal="+ accountBalance + " alertDate"+ alertDate + " diff" + getDifferenceDays(alertDate, new Date()));
+                if (accountBalance.compareTo(zeroBalance) <= 0) {
+                    if (_lockAccountEnforcement && (lockable == 1)) {
+                        if (account.getType() == Account.ACCOUNT_TYPE_NORMAL) {
+                            lockAccount(account.getId());
+                        }
+                    }
+                    if (alertDate == null || (balanceDate.after(alertDate) && getDifferenceDays(alertDate, new Date()) > 1)) {
+                        deferredQuotaEmailList.add(new DeferredQuotaEmail(account, quotaAccount, QuotaConfig.QuotaEmailTemplateTypes.QUOTA_EMPTY));
+                    }
+                } else if (accountBalance.compareTo(thresholdBalance) <= 0) {
+                    if (alertDate == null || (balanceDate.after(alertDate) && getDifferenceDays(alertDate, new Date()) > 1)) {
+                        deferredQuotaEmailList.add(new DeferredQuotaEmail(account, quotaAccount, QuotaConfig.QuotaEmailTemplateTypes.QUOTA_LOW));
+                    }
+                }
+            }
+        }
+
+        for (DeferredQuotaEmail emailToBeSent : deferredQuotaEmailList) {
+            s_logger.debug("Attempting to send quota alert email to users of account: " + emailToBeSent.getAccount().getAccountName());
+            sendQuotaAlert(emailToBeSent);
+        }
+    }
+
+    private void sendQuotaAlert(DeferredQuotaEmail emailToBeSent) {
+        final AccountVO account = emailToBeSent.getAccount();
+        final BigDecimal balance = emailToBeSent.getQuotaBalance();
+        final BigDecimal usage = emailToBeSent.getQuotaUsage();
+        final QuotaConfig.QuotaEmailTemplateTypes emailType = emailToBeSent.getEmailTemplateType();
+
+        final List<QuotaEmailTemplatesVO> emailTemplates = _quotaEmailTemplateDao.listAllQuotaEmailTemplates(emailType.toString());
+        if (emailTemplates != null && emailTemplates.get(0) != null) {
+            final QuotaEmailTemplatesVO emailTemplate = emailTemplates.get(0);
+
+            final DomainVO accountDomain = _domainDao.findByIdIncludingRemoved(account.getDomainId());
+            final List<UserVO> usersInAccount = _userDao.listByAccount(account.getId());
+
+            String userNames = "";
```

Comment:
Why not use a ``List`` rather than a ``String`` and use ``com.google.command.base.Joiner`` to create a comma-separated string when needed?